### PR TITLE
Concatenate contiguous strings

### LINKED ISSUE #16820: Convert to Interpolated String inserts an extra space in the string from newlines
**Version Used**: 15.0.0-RC.3+26127.0

```C#
var s = "A" + 1 + "B"
      + "C";

```

The Convert to Interpolated String refactoring produces

```C#
var s = $"A{1}B C";
```

The space between `B` and `C` should not be there.
This a subtle and easily-overlooked change in runtime behavior; this is quite bad.

### PR DESCRIPTION
**Pull request by jmarolf**
Concatenate contiguous strings when converting from string concatenation to interpolated string
fixes #16820

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/ConvertToInterpolatedString/ConvertConcatenationToInterpolatedStringTests.cs [hunk 1]
```diff
@@ -322,5 +322,26 @@ void M()
     }
 }");
         }
+
+        [WorkItem(16820, "https://github.com/dotnet/roslyn/issues/16820")]
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertToInterpolatedString)]
+        public async Task TestWithMultipleStringConcatinations()
```
Comment (DustinCampbell) [inline src/EditorFeatures/CSharpTest/ConvertToInterpolatedString/ConvertConcatenationToInterpolatedStringTests.cs]: Maybe add a test mixing normal and verbatim string literals?

**Comment by jmarolf** [comment]
@dotnet/roslyn-ide please review

File: src/EditorFeatures/CSharpTest/ConvertToInterpolatedString/ConvertConcatenationToInterpolatedStringTests.cs (hunk 1 above)
Comment (jmarolf) [inline src/EditorFeatures/CSharpTest/ConvertToInterpolatedString/ConvertConcatenationToInterpolatedStringTests.cs]: We don't trigger if we detect this case.  I've added a test to make sure this remains the case.

File: src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs [hunk 2]
```diff
@@ -112,22 +112,45 @@ private Task<Document> UpdateDocumentAsync(Document document, SyntaxNode root, S
...
             {
-                if (syntaxFacts.IsStringLiteralExpression(piece))
+                if (syntaxFacts.IsStringLiteralExpression(pieces[i]))
                 {
-                    var text = piece.GetFirstToken().Text;
+                    var text = pieces[i].GetFirstToken().Text;
                     var textWithoutQuotes = GetTextWithoutQuotes(text, isVerbatimStringLiteral);
-                    content.Add(generator.InterpolatedStringText(
-                        generator.InterpolatedStringTextToken(textWithoutQuotes)));
+                    if (i-1 == previousIndex)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs]: spacing around -

File: src/EditorFeatures/CSharpTest/ConvertToInterpolatedString/ConvertConcatenationToInterpolatedStringTests.cs [hunk 3]
```diff
@@ -322,5 +322,40 @@ void M()
...
+        {
+            await TestAsync(
+@"public class C
+{
+    void M()
+    {
+        var v = ""A"" + 1 + [||]""B"" + ""C"";
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/ConvertToInterpolatedString/ConvertConcatenationToInterpolatedStringTests.cs]: test with more strange stuff.  like "A" + "B" + 1   and    "A" + 1 + "B" + "C" + "D"

File: src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs [hunk 4]
```diff
@@ -112,22 +111,45 @@ private Task<Document> UpdateDocumentAsync(Document document, SyntaxNode root, S
...
+                    if (i - 1 == previousIndex)
+                    {
+                        // Last part we added to the content list was also an interpolated-string-text-node.
+                        // We need to combine these as the API for creating an interpolated strings
+                        // does not expect to be given a list containing non-contiguous string nodes.
+                        // Essentially if we combine '"A" + 1 + "B" + "C"' into '$"A{1}BC"' it must be:
+                        //      {InterpolatedStringText}{Interpolation}{InterpolatedStringText}
+                        // not:
+                        //      {InterpolatedStringText}{Interpolation}{InterpolatedStringText}{InterpolatedStringText}
+                        var interpolatedStringTextNode = content[insertionIndex];
```
Comment (Pilchie) [inline src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs]: I'm not thrilled with trying to do the mental gymnastics to prove that this never gets called with -1 or after the end of content.  Convince me (and future readers) that this is safe.

File: src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs [hunk 5]
```diff
@@ -105,22 +105,20 @@ private Task<Document> UpdateDocumentAsync(Document document, SyntaxNode root, S
...
 
             var content = new List<SyntaxNode>(pieces.Count);
-            var previousIndex = -2;
-            var insertionIndex = -1;
-            for (var i = 0; i < pieces.Count; i++)
+            var previousContentWasStringLiteralExpression = false;
+            foreach (var piece in pieces)
             {
-                if (syntaxFacts.IsStringLiteralExpression(pieces[i]))
+                if (syntaxFacts.IsStringLiteralExpression(piece))
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs]: small possible tweaks:

```
var currentIsStringLiteral = syntaxFacts.IsStringLiteralExpression();
...at very bottom of loop
previousWasStringLiteral = currentIsStringLiteral
```

File: src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs [hunk 6]
```diff
@@ -129,30 +127,39 @@ private Task<Document> UpdateDocumentAsync(Document document, SyntaxNode root, S
                         //      {InterpolatedStringText}{Interpolation}{InterpolatedStringText}
                         // not:
                         //      {InterpolatedStringText}{Interpolation}{InterpolatedStringText}{InterpolatedStringText}
-                        var interpolatedStringTextNode = content[insertionIndex];
-                        var existingText = interpolatedStringTextNode.GetFirstToken().Text;
-                        var newText = existingText + textWithoutQuotes;
-                        content[insertionIndex] = generator.InterpolatedStringText(generator.InterpolatedStringTextToken(newText));
+                        var existingingInterpolatedStringTextNode = content.Last();
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs]: existing ing ing ing

**Comment by CyrusNajmabadi** [comment]
Man, this is much better 👍

**Comment by jmarolf** [comment]
test windows_debug_unit32_prtest please

**Pull request merged**